src/lib.rs

```diff
@@ -198,67 +287,200 @@ impl IRCompilerFactory {
         unsafe {
             self.funcs
-                .IRCompilerSetEntryPointName(self.me, new_name.as_ptr())
+                .IRCompilerSetEntryPointName(self.me.as_ptr(), new_name.as_ptr())
         }
     }
 
+    #[doc(alias = "IRCompilerAllocCompileAndLink")]
     pub fn alloc_compile_and_link(
         &mut self,
         entry_point: &CStr,
         input: &IRObject,
-    ) -> Result<IRObject, Box<dyn std::error::Error>> {
+    ) -> Result<IRObject, CompilerError> {
         let mut error = std::ptr::null_mut();
 
-        let v = unsafe {
+        let v = NonNull::new(unsafe {
             self.funcs.IRCompilerAllocCompileAndLink(
-                self.me,
+                self.me.as_ptr(),
                 entry_point.as_ptr(),
-                input.me,
+                input.me.as_ptr(),
                 &mut error,
             )
-        };
+        });
 
         if error.is_null() {
             Ok(IRObject {
-                me: v,
+                me: v.unwrap(),
                 funcs: input.funcs.clone(),
             })
         } else {
-            panic!("{:?}", error);
+            let error = IRError::from_ptr(error, self.funcs.clone());
+            Err(CompilerError(error.code()))
+        }
+    }
+}
+
+struct IRError {
+    me: NonNull<bindings::IRError>,
+    funcs: Arc<bindings::metal_irconverter>,
+}
+
+impl Drop for IRError {
+    #[doc(alias = "IRErrorDestroy")]
+    fn drop(&mut self) {
+        unsafe { self.funcs.IRErrorDestroy(self.me.as_ptr()) };
+    }
+}
+
+impl IRError {
+    pub fn from_ptr(ptr: *mut bindings::IRError, funcs: Arc<bindings::metal_irconverter>) -> Self {
+        Self {
+            me: NonNull::new(ptr).unwrap(),
+            funcs,
+        }
+    }
+
+    #[doc(alias = "IRErrorGetCode")]
+    pub fn code(&self) -> IRErrorCode {
+        let code = unsafe { self.funcs.IRErrorGetCode(self.me.as_ptr()) };
+        match code {
+            0 => IRErrorCode::NoError,
+            1 => IRErrorCode::ShaderRequiresRootSignature,
+            2 => IRErrorCode::UnrecognizedRootSignatureDescriptor,
+            3 => IRErrorCode::UnrecognizedParameterTypeInRootSignature,
+            4 => IRErrorCode::ResourceNotReferencedByRootSignature,
+            5 => IRErrorCode::ShaderIncompatibleWithDualSourceBlending,
+            6 => IRErrorCode::UnsupportedWaveSize,
+            7 => IRErrorCode::UnsupportedInstruction,
+            8 => IRErrorCode::CompilationError,
+            9 => IRErrorCode::FailedToSynthesizeStageInFunction,
+            10 => IRErrorCode::FailedToSynthesizeStreamOutFunction,
+            11 => IRErrorCode::FailedToSynthesizeIndirectIntersectionFunction,
+            12 => IRErrorCode::UnableToVerifyModule,
+            13 => IRErrorCode::UnableToLinkModule,
+            14 => IRErrorCode::UnrecognizedDXILHeader,
+            15 => IRErrorCode::InvalidRaytracingAttribute,
+            16 => IRErrorCode::Unknown,
```

Comment:
Never copy-paste numbers from bindings, this'll get out of date as soon as upstream changes something. Let's switch to:

```rust
x if x == IRErrorCode::Unknown as u32 => ...,
```

Even better if we can have `num_enum` via `bindgen`.